File: week9/김정호/N으로 표현.java
class Solution {

    static int min = 9;

    public int solution(int N, int number) {
        dfs(0, N, number, 0);

        if (min == 9) return -1;
        return min;
    }

    public void dfs(int depth, int N, int number, int current) {
        if (depth > 8) {
            return;
        }

        if (number == current) {
            min = Math.min(depth, min);
            return;
        }

        int temp = 0;

        for (int i = 0; i < 8; i++) {
            if (depth + i < 8) {
                temp = temp * 10 + N; // 자리수 늘리는 부분 (5, 55, 555)
                dfs(depth + i + 1, N, number, current + temp);
                dfs(depth + i + 1, N, number, current - temp);
                dfs(depth + i + 1, N, number, current / temp);
                dfs(depth + i + 1, N, number, current * temp);
            }
        }
    }
}

File: week9/한지희/N으로_표현.kt
import kotlin.math.min

class ExpressInN {
    var target = 0
    var n = 0
    var answer = Int.MAX_VALUE

    fun solution(N: Int, number: Int): Int {
        n = N
        target = number

        logic(0, 0)

        if (answer > 8) answer = -1

        return answer
    }

    private fun logic(count: Int, current: Int) {
        if (count > 8) return
        if (current == target) {
            answer = min(answer, count)
            return
        }

        var value = n
        for (i in 0 until (8 - count)) {
            logic(count + 1 + i, current + value)
            logic(count + 1 + i, current - value)
            logic(count + 1 + i, current * value)
            logic(count + 1 + i, current / value)
            value = ("$value" + "$n").toInt()
        }
    }
}